Publication quality plots (#110)
yt-project/yt-4.0-paper: content/45.user_friendliness.md

## User-Friendliness

Several different entrypoints exist for research codes.
For instance, an ontology of "end-users" in science has developed that emphasizes that typically "library" codes are only used implicitly by researchers.
In this framing, library codes primarily interface with application codes, rather than with researchers.

> We have taken a slightly different approach with the design and development of yt, as it exists in a middle-ground as a library used *as an application* within the scripting language python and also as a library for more complex analysis of data.

Reviewer: ```suggestion
We have taken a slightly different approach with the design and development of yt, as it exists in a middle-ground as a library used *as an application* within the scripting language Python and also as a library for more complex analysis of data.
```

what part of yt would you describe as an application ? I would think of the CLI as such, but is there anything else ?

Author: So, my phrasing was kind of deliberate -- it's a library that people use as an application.  Basically, a lot of the stuff they do is what amounts to GUI operations mapped onto a reasonably small API.  I'm not saying it's an application, but I would argue that it's used similarly to one.

Reviewer: I see, then I have no issues with that

As an outgrowth of this, we have taken particular care with the "public-facing" API of yt.
We have attempted to abstract the API enough from the data structures yt uses internally that it is useful without detailed knowledge of yt's internals.
By the same token, we have also attempted to provide low-level access to data, and to make as many of those methods accessible and usable as well.

> In addition to API considerations, there are three areas that we note have additional care paid to the generation of figures, useful error messages, and to integration with the omnipresent Jupyter environment.

Reviewer: ```suggestion
In addition to API considerations, there are three areas that we note have additional care paid to the generation of figures, helpful error messages, and to integration with the omnipresent Jupyter environment.
```

### Publication-Ready Figures

> Matplotlib @doi:10.5281/zenodo.7570264 is a fully-featured mechanism for generating figures, with an incredible array of options to customize formatting, appearance, rendering of fonts and glyphs, sleection of colorbars, calculation of tick locations, and appropriate bounds.

Reviewer: ```suggestion
Matplotlib @doi:10.5281/zenodo.7570264 is a fully-featured mechanism for generating figures, with an incredible array of options to customize formatting, appearance, rendering of fonts and glyphs, selection of colorbars, calculation of tick locations, and appropriate bounds.
```

This degree of flexibility provides extremely fine-grained control over the appearance of figures for publication, and Matplotlib provides a large number of output formats for even the most discerning of journals.

> yt utilizes matplotlib as its primary rendering engine for visualizations; while 3D renderings (such as those described in @sec:vr) typically generate raw pixel buffers that are saved as images directly, most other visualization functionality in yt relies on matplotlib for generating images that are saved to disk.

Reviewer: ```suggestion
yt utilizes Matplotlib as its primary rendering engine for visualizations; while 3D renderings (such as those described in @sec:vr) typically generate raw pixel buffers that are saved as images directly, most other visualization functionality in yt relies on Matplotlib for generating images that are saved to disk.
```

> Slices and projections are prepared as variable-resolution images that are pixelized (@sec:pixelization) into fixed resolution buffers, then provided to matplotlib in the `imshow` command.

Reviewer: ```suggestion
Slices and projections are prepared as variable-resolution images that are pixelized (@sec:pixelization) into fixed resolution buffers, then provided to Matplotlib in the `imshow` function.
```

> Phase plots utilize the `pcolormesh` operation.

Reviewer: ```suggestion
Phase plots utilize the `pcolormesh` function.
```

I never realised that. Just curious, why not imshow ?

> All of these internal plots utilize matplotlib's "object-oriented" interface, wherein `Figure` and `Axes` objects are created directly.

Reviewer: ```suggestion
All of these internal plots utilize Matplotlib's "object-oriented" interface, wherein `Figure` and `Axes` objects are created directly.
```

yt then manages the internal state of these objects, and provides high-level access to operations that are aware of the nature of the visualizations; for instance, operations that zoom and pan, and that are unit-aware.
In the case of 2D spatial images, these are all generated as subclasses of a `PlotWindow` object, named such because it functions as a "window" onto the data, including automatically generating and managing multiple fields simultaneously.

> Within the `PlotWindow` object, we set up two levels of data invalidation; when a characteristic of the data is changed (i.e., the viewing window, the resolution) then the *data* is said to be invalidated.

Reviewer: I've always found the term "data" to be pretty confusing in this context: it's not the simulation data itself that's invalidated, rather the _image_ data. I think this should be clarified

> When a characteristic of the visualization is changed (the label, the scaling of the colorbar, the colormap) then the *plot* is said to be invalidated.

Reviewer: there's also a third layer consisting in `@(in)validate_figure` decorators, but it never really clicked for me so I can't tell if it's worth discussing here.

Author: I forgot we had those.  Maybe they were for the case that we need to change dpi or something.

> This allows yt to minimize the overhead of conducting potentially expensive operations on the underlying data, while maximizing its utilization of matplotlib for generation of images.

Reviewer: ```suggestion
This allows yt to minimize the overhead of conducting potentially expensive operations on the underlying data, while maximizing its utilization of Matplotlib for generation of images.
```

> The image buffers generated are also always available through the API, so in the not-uncommon case that researchers wish to take the genrated images and perform detailed manipulations or plot layouts, they are able to do so, using the matplotlib API they may already be familiar with.

Reviewer: ```suggestion
The image buffers generated are also always available through the API, so in the common case that researchers wish to take the genrated images and perform detailed manipulations or plot layouts, they are able to do so, using the Matplotlib API they may already be familiar with.
```

Author: ![not-undelicious](https://user-images.githubusercontent.com/89019/216096287-f8466ca3-f125-48a0-bc7e-c09bd19cd964.jpg)
![food-delicious](https://user-images.githubusercontent.com/89019/216096300-8c7a42ab-1e79-438f-bdbc-5bc59315dc52.jpg)

Additional data-aware operations include overplotting locations of gravitationally-bound clumps, contour plots of variable-resolution data, boundaries of data structures (such as grids in patch-based grid datasets, cells from patch and octree datasets, particle locations), streamlines, line integral convolution, and coordinate-aware annotations.

All of these figure construction and modification methods apply to cartesian ($x$, $y$, $z$) datasets, but yt also supports them for curvilinear data such as that organized spherically ($r$, $\theta$, $\phi$) and cylindrically ($r$, $z$, $\theta$).

By focusing on the utilizing low-level yt operations to facilitate high-level interactions, we can embed best practices for accessibility and visualizations at as many levels as possible.
For instance, yt provides suggested colormaps for specific fields, will apply heuristics to determine the appropriate scaling (such as symmetric logarithm) and utilize the appropriate labels and units to the data.

> This allows users of yt to access both the expansive functionality of matplotlib as well as the more domain-specific tasks available in yt.

Reviewer: ```suggestion
This allows users of yt to access both the expansive functionality of Matplotlib as well as the more domain-specific tasks available in yt.
```
